README.txt
JSON-RPC 2.0 Server

Copyright (c) Vladimir Dzhuvinov, 2010 - 2012

README

Simple server framework for processing JSON-RPC 2.0 requests and notifications.


Usage:

	1. Implement request and / or notification handlers for the various 
	   expected JSON-RPC 2.0 messages. A handler may process one or more 
	   request/notification methods (identified by method name).
	   
	2. Create a new Dispatcher and register the handlers with it.
	
	3. Pass the received JSON-RPC 2.0 requests and notifications to the 
	   appropriate Dispatcher.dispatch(...) method, then, if the message is 
	   a request, pass back to the client the resulting JSON-RPC 2.0 
	   response. 


Package requirements:

	* Java 1.5 or later
	
	* JSON-RPC 2.0 Base for creating and representing JSON-RPC 2.0 messages.
	  This package is included in the lib/ directory. Online documentation 
	  is at http://software.dzhuvinov.com/json-rpc-2.0-base.html.
	
	* Java Servlet API for constructing MessageContext objects from HTTP 
	  servlet requests. 


Package content:

	README.txt                    This file.
	
	LICENSE.txt                   The software license.
	
	CHANGELOG.txt                 The change log.
	
	jsonrpc2-server-{version}.jar JAR file containing the compiled package
	                              files.
	
	Example.java                  Example program.
	
	javadoc/                      The Java Docs for this package.
	
	build.xml                     Apache Ant build file.
	
	lib/                          The package dependencies.
	
	src/                          The full source code for this package.



For complete JSON-RPC 2.0 Server documentation, examples and updates visit

	http://software.dzhuvinov.com/json-rpc-2.0-server.html


The JSON-RPC 2.0 specification and user group forum can be found at

        http://groups.google.com/group/json-rpc

[EOF]


CHANGELOG.txt
CHANGE LOG

JsonRpc2-Server, Copyright (c) Vladimir Dzhuvinov, 2010 - 2011


version 1.0 (2010-05-18)
	* First release.

version 1.1 (2010-10-10)
	* Renames class RequestInfo to RequestContext.

version 1.2 (2010-11-23)
	* Renames package.
	* Removes dependency on org.apache.commons.lang.

version 1.3 (2011-03-05)
	* Major API naming changes.
	* Adds support for handling JSON-RPC 2.0 notifications.
	* Allows for undefined (null) message context arguments in
	  Dispatcher.dispatch() and {Request|Notification}Handler.process() 
	  methods.

version 1.4 (2011-07-11)
	* Raises minimal JSON-RPC 2.0 Base requirement to version 1.19.
	* Adds support for reporting request processing time by adding a 
	  non-standard "_procTime" attribute to the JSON-RPC 2.0 response.

version 1.4.1 (2011-07-15)
	* Renames the non-standard processing time attribute to "xProcTime".
	* Documentation improvements.

version 1.5 (2012-02-19)
	* Fixes a bug in MessageContext getter method for 
	  java.security.Principal.
	* MessageContext allows storing of multiple java.security.Principal
	  identities.
	* Updates JSON-RPC 2.0 Base JAR to 1.25 (JSON Smart 1.1).

version 1.5.1 (2012-04-03)
	* Updates JSON-RPC 2.0 Base JAR to 1.25.1 (JSON Smart 1.1.1).

version 1.5.2 (2012-07-15)
	* Add pom.xml.


Example.java
import com.thetransactioncompany.jsonrpc2.*;
import com.thetransactioncompany.jsonrpc2.server.*;

import java.text.*;
import java.util.*;


/**
 * Demonstration of the JSON-RPC 2.0 Server framework usage. The request
 * handlers are implemented as static nested classes for convenience, but in 
 * real life applications may be defined as regular classes within their old 
 * source files.
 *
 * @author Vladimir Dzhuvinov
 * @version 2011-03-05
 */ 
public class Example {


	// Implements a handler for an "echo" JSON-RPC method
	public static class EchoHandler implements RequestHandler {
	
	
		// Reports the method names of the handled requests
		public String[] handledRequests() {
		
			return new String[]{"echo"};
		}
		
		
		// Processes the requests
		public JSONRPC2Response process(JSONRPC2Request req, MessageContext ctx) {
			
			if (req.getMethod().equals("echo")) {
				
				// Echo first parameter
				
				List params = (List)req.getParams();
			
				Object input = params.get(0);
			
				return new JSONRPC2Response(input, req.getID());
			}
			else {
				// Method name not supported
				
				return new JSONRPC2Response(JSONRPC2Error.METHOD_NOT_FOUND, req.getID());
			}
		}
	}
	
	
	// Implements a handler for "getDate" and "getTime" JSON-RPC methods
	// that return the current date and time
	public static class DateTimeHandler implements RequestHandler {
	
	
		// Reports the method names of the handled requests
		public String[] handledRequests() {
		
			return new String[]{"getDate", "getTime"};
		}
		
		
		// Processes the requests
		public JSONRPC2Response process(JSONRPC2Request req, MessageContext ctx) {
		
			if (req.getMethod().equals("getDate")) {
			
				DateFormat df = DateFormat.getDateInstance();
				
				String date = df.format(new Date());
				
				return new JSONRPC2Response(date, req.getID());
			}
			else if (req.getMethod().equals("getTime")) {
			
				DateFormat df = DateFormat.getTimeInstance();
				
				String time = df.format(new Date());
				
				return new JSONRPC2Response(time, req.getID());
			}
			else {
			
				// Method name not supported
				
				return new JSONRPC2Response(JSONRPC2Error.METHOD_NOT_FOUND, req.getID());
			}
		}
	}


	public static void main(String[] args) {
	
		// Create a new JSON-RPC 2.0 request dispatcher
		Dispatcher dispatcher =  new Dispatcher();
		
		
		// Register the "echo", "getDate" and "getTime" handlers with it
		dispatcher.register(new EchoHandler());
		dispatcher.register(new DateTimeHandler());
		
		// Simulate an "echo" JSON-RPC 2.0 request
		List echoParam = new LinkedList();
		echoParam.add("Hello world!");
		
		JSONRPC2Request req = new JSONRPC2Request("echo", echoParam, "req-id-01");
		System.out.println("Request: \n" + req);
		
		JSONRPC2Response resp = dispatcher.dispatch(req, null);
		System.out.println("Response: \n" + resp);
		
		
		// Simulate a "getDate" JSON-RPC 2.0 request
		req = new JSONRPC2Request("getDate", "req-id-02");
		System.out.println("Request: \n" + req);
		
		resp = dispatcher.dispatch(req, null);
		System.out.println("Response: \n" + resp);
		
		
		// Simulate a "getTime" JSON-RPC 2.0 request
		req = new JSONRPC2Request("getTime", "req-id-03");
		System.out.println("Request: \n" + req);
		
		resp = dispatcher.dispatch(req, null);
		System.out.println("Response: \n" + resp);
	}
}


src/main/java/com/thetransactioncompany/jsonrpc2/server/package-info.java
/** 
 * Simple server framework for processing JSON-RPC 2.0 requests and
 * notifications.
 *
 * <p>Usage:
 *
 * <ol>
 *     <li>Implement {@link com.thetransactioncompany.jsonrpc2.server.RequestHandler request}
 *         and / or {@link com.thetransactioncompany.jsonrpc2.server.NotificationHandler notification}
 *         handlers for the various expected JSON-RPC 2.0 messages. A handler
 *         may process one or more request/notification methods (identified by 
 *         method name).
 *     <li>Create a new {@link com.thetransactioncompany.jsonrpc2.server.Dispatcher}
 *         and register the handlers with it.
 *     <li>Pass the received JSON-RPC 2.0 requests and notifications to the 
 *         appropriate {@code Dispatcher.dispatch(...)} method, then, if the 
 *         message is a request, pass the resulting JSON-RPC 2.0 response back
 *         to the client.
 * </ol>
 *
 * <p>Direct package dependencies:
 *
 * <ul>
 *     <li><b><a href="http://software.dzhuvinov.com/json-rpc-2.0-base.html">JSON-RPC 2.0 Base</a></b> 
 *        [<i>com.thetransactioncompany.jsonrpc2</i>] to construct and represent 
 *         JSON-RPC 2.0 messages.
 *     <li><b>Java Servlet API</b> [<i>javax.servlet.http</i>] for constructing 
 *         {@link com.thetransactioncompany.jsonrpc2.server.MessageContext}
 *         objects from HTTP servlet requests.
 * </ul>
 *
 *
 * @author Vladimir Dzhuvinov
 * @version 1.5.1 (2012-02-20)
 */
package com.thetransactioncompany.jsonrpc2.server;


src/main/java/com/thetransactioncompany/jsonrpc2/server/MessageContext.java
package com.thetransactioncompany.jsonrpc2.server;


import java.security.Principal;
import java.security.cert.X509Certificate;
import javax.servlet.http.HttpServletRequest;


/**
 * Context information about JSON-RPC 2.0 request and notification messages.
 *
 * <ul>
 *     <li>The client's host name.
 *     <li>The client's IP address.
 *     <li>Whether the request / notification was transmitted securely (e.g. 
 *         via HTTPS).
 *     <li>The client principal(s) (user), if authenticated.
 * </ul>
 *
 * @author Vladimir Dzhuvinov
 * @version 1.5.1 (2012-02-20)
 */
public class MessageContext {


	/** 
	 * The client hostname, {@code null} if none was specified.
	 */
	private String clientHostName = null;

	
	/** 
	 * The client IP address, {@code null} if none was specified.
	 */
	private String clientInetAddress = null;

	
	/** 
	 * Indicates whether the request was received over a secure channel
	 * (typically HTTPS). 
	 */
	private boolean secure = false;
	
	
	/**
	 * The authenticated client principals, {@code null} if none were 
	 * specified.
	 */
	private Principal[] principals = null;
	
	
	/**
	 * Minimal implementation of the {@link java.security.Principal} 
	 * interface.
	 */
	public class BasicPrincipal implements Principal {
	
		/**
		 * The principal name.
		 */
		private String name;
	
	
		/**
		 * Creates a new principal.
		 *
		 * @param name The principal name, must not be {@code null} or
		 *             empty string.
		 *
		 * @throws IllegalArgumentException On a {@code null} or empty 
		 *                                  principal name.
		 */
		public BasicPrincipal(final String name) {
		
			if (name == null || name.trim().isEmpty())
				throw new IllegalArgumentException("The principal name must be defined");
		
			this.name = name;
		}
	
	
		/**
		 * Checks for equality.
		 *
		 * @param another The object to compare to.
		 */
		public boolean equals(final Object another) {
		
			if (another instanceof Principal && ((Principal)another).getName().equals(this.getName()))
				return true;
			else
				return false;
		}
		
		
		/**
		 * Returns a hash code for this principal.
		 *
		 * @return The hash code.
		 */
		public int hashCode() {
		
			return getName().hashCode();
		}
		
		
		/**
		 * Returns the principal name.
		 *
		 * @return The principal name.
		 */
		public String getName() {
			
			return name;
		}
	}
	
	
	/**
	 * Creates a new JSON-RPC 2.0 request / notification context.
	 *
	 * @param clientHostName    The client host name, {@code null} if 
	 *                          unknown.
	 * @param clientInetAddress The client IP address, {@code null} if 
	 *                          unknown.
	 * @param secure            Specifies a request received over HTTPS.
	 * @param principalName     Specifies the authenticated client principle
	 *                          name, {@code null} if unknown. The name must
	 *                          not be an empty or blank string.
	 */
	public MessageContext(final String clientHostName, 
	                      final String clientInetAddress, 
			      final boolean secure, 
			      final String principalName) {
	
		this.clientHostName = clientHostName;
		this.clientInetAddress = clientInetAddress;
		this.secure = secure;
		
		if (principalName != null) {
			principals = new Principal[1];
			principals[0] = new BasicPrincipal(principalName);
		}
	}
	
	
	/**
	 * Creates a new JSON-RPC 2.0 request / notification context.
	 *
	 * @param clientHostName    The client host name, {@code null} if 
	 *                          unknown.
	 * @param clientInetAddress The client IP address, {@code null} if 
	 *                          unknown.
	 * @param secure            Specifies a request received over HTTPS.
	 * @param principalNames    Specifies the authenticated client principle
	 *                          names, {@code null} if unknown. The names
	 *                          must not be an empty or blank string.
	 */
	public MessageContext(final String clientHostName, 
	                      final String clientInetAddress, 
			      final boolean secure, 
			      final String[] principalNames) {
	
		this.clientHostName = clientHostName;
		this.clientInetAddress = clientInetAddress;
		this.secure = secure;
		
		if (principalNames != null) {
			principals = new Principal[principalNames.length];
			
			for (int i=0; i < principals.length; i++)
				principals[0] = new BasicPrincipal(principalNames[i]);
		}
	}
	
	
	/**
	 * Creates a new JSON-RPC 2.0 request / notification context. No 
	 * authenticated client principal is specified.
	 *
	 * @param clientHostName    The client host name, {@code null} if 
	 *                          unknown.
	 * @param clientInetAddress The client IP address, {@code null} if 
	 *                          unknown.
	 * @param secure            Specifies a request received over HTTPS.
	 */
	public MessageContext(final String clientHostName, 
	                      final String clientInetAddress, 
			      final boolean secure) {
	
		this.clientHostName = clientHostName;
		this.clientInetAddress = clientInetAddress;
		this.secure = secure;
	}
	
	
	/**
	 * Creates a new JSON-RPC 2.0 request / notification context. Indicates 
	 * an insecure transport (plain HTTP) and no authenticated client 
	 * principal.
	 *
	 * @param clientHostName    The client host name, {@code null} if 
	 *                          unknown.
	 * @param clientInetAddress The client IP address, {@code null} if 
	 *                          unknown.
	 */
	public MessageContext(final String clientHostName, 
	                      final String clientInetAddress) {
	
		this.clientHostName = clientHostName;
		this.clientInetAddress = clientInetAddress;
		this.secure = false;
	}
	
	
	/**
	 * Creates a new JSON-RPC 2.0 request / notification context from the
	 * specified HTTP request.
	 *
	 * @param httpRequest The HTTP request.
	 */
	public MessageContext(final HttpServletRequest httpRequest) {
	
		clientInetAddress = httpRequest.getRemoteAddr();
	
		clientHostName = httpRequest.getRemoteHost();
		
		if (clientHostName != null && clientHostName.equals(clientInetAddress))
			clientHostName = null; // not resolved actually
		
		secure = httpRequest.isSecure();

		X509Certificate[] certs = (X509Certificate[])httpRequest.getAttribute("javax.servlet.request.X509Certificate");

		if (certs != null && certs.length > 0) {	
			
			principals = new Principal[certs.length];
			
			for (int i=0; i < principals.length; i++)
				principals[i] = certs[i].getSubjectX500Principal();
		}
	}
	
	
	/**
	 * Gets the host name of the client that sent the request / 
	 * notification.
	 *
	 * @return The client host name, {@code null} if unknown.
	 */
	public String getClientHostName() {
	
		return clientHostName;
	}
	
	
	/**
	 * Gets the IP address of the client that sent the request /
	 * notification.
	 *
	 * @return The client IP address, {@code null} if unknown.
	 */
	public String getClientInetAddress() {
		
		return clientInetAddress;
	}
	 

	/**
	 * Indicates whether the request / notification was received over a 
	 * secure HTTPS connection.
	 *
	 * @return {@code true} If the request was received over HTTPS, 
	 *         {@code false} if it was received over plain HTTP.
	 */
	public boolean isSecure() {
	
		return secure;
	}
	
	
	/**
	 * Returns the first authenticated client principal, {@code null} if 
	 * none.
	 *
	 * @return The first client principal, {@code null} if none.
	 */
	public Principal getPrincipal() {
	
		if (principals != null)
			return principals[0];
		else
			return null;
	}
	
	
	/**
	 * Returns the authenticated client principals, {@code null} if 
	 * none.
	 *
	 * @return The client principals, {@code null} if none.
	 */
	public Principal[] getPrincipals() {
	
		return principals;
	}
	
	
	/**
	 * Returns the first authenticated client principal name, {@code null} 
	 * if none.
	 *
	 * @return The frist client principal name, {@code null} if none.
	 */
	public String getPrincipalName() {
	
		if (principals != null)
			return principals[0].getName();
		else
			return null;
	}
	
	
	/**
	 * Returns the authenticated client principal names, {@code null} 
	 * if none.
	 *
	 * @return The client principal names, {@code null} if none.
	 */
	public String[] getPrincipalNames() {
	
		String[] names = new String[principals.length];
		
		for (int i=0; i < names.length; i++)
			names[i] = principals[i].getName();
		
		return names;
	}
}


src/main/java/com/thetransactioncompany/jsonrpc2/server/RequestHandler.java
package com.thetransactioncompany.jsonrpc2.server;


import com.thetransactioncompany.jsonrpc2.*;


/**
 * Interface for handling JSON-RPC 2.0 requests.
 *
 * @author Vladimir Dzhuvinov
 * @version 1.5.1 (2011-03-05)
 */
public interface RequestHandler {

	
	/**
	 * Gets the JSON-RPC 2.0 request method names that this handler 
	 * processes.
	 *
	 * @return The method names of the served JSON-RPC 2.0 requests.
	 */
	public String[] handledRequests();
	
	
	/**
	 * Processes a JSON-RPC 2.0 request.
	 *
	 * @param request    A valid JSON-RPC 2.0 request instance.
	 * @param requestCtx Context information about the request, may be 
	 *                   {@code null} if undefined.
	 *
	 * @return The resulting JSON-RPC 2.0 response.
	 */
	public JSONRPC2Response process(final JSONRPC2Request request, final MessageContext requestCtx);

}


src/main/java/com/thetransactioncompany/jsonrpc2/server/NotificationHandler.java
package com.thetransactioncompany.jsonrpc2.server;


import com.thetransactioncompany.jsonrpc2.*;


/**
 * Interface for handling JSON-RPC 2.0 notifications.
 *
 * @author Vladimir Dzhuvinov
 * @version 1.5.1 (2011-03-05)
 */
public interface NotificationHandler {

	
	/**
	 * Gets the JSON-RPC 2.0 notification method names that this handler 
	 * processes.
	 *
	 * @return The method names of the served JSON-RPC 2.0 notifications.
	 */
	public String[] handledNotifications();
	
	
	/**
	 * Processes a JSON-RPC 2.0 notification.
	 *
	 * @param notification    A valid JSON-RPC 2.0 notification instance.
	 * @param notificationCtx Context information about the notification,
	 *                        may be {@code null} if undefined.
	 */
	public void process(final JSONRPC2Notification notification, final MessageContext notificationCtx);

}


src/main/java/com/thetransactioncompany/jsonrpc2/server/Dispatcher.java
package com.thetransactioncompany.jsonrpc2.server;


import java.util.*;

import com.thetransactioncompany.jsonrpc2.*;


/**
 * Dispatcher for JSON-RPC 2.0 requests and notifications.
 *
 * <p>Use the {@code register()} methods to add a request or notification
 * handler for an RPC method.
 *
 * <p>Use the {@code dispatch()} methods to have an incoming request or
 * notification processed by the matching handler.
 *
 * <p>The {@code reportProcTime()} method enables reporting of request 
 * processing time (in microseconds) by appending a non-standard "xProcTime" 
 * attribute to the resulting JSON-RPC 2.0 response message.
 *
 * <p>Example:
 *
 * <pre>
 * { 
 *   "result"    : "xyz",
 *   "id"        : 1,
 *   "jsonrpc"   : "2.0",
 *   "xProcTime" : "189 us"
 * }
 * </pre>
 *
 * @author Vladimir Dzhuvinov
 * @version 1.5.1 (2011-07-15)
 */
public class Dispatcher {
	
	
	/** 
	 * Hashtable of request name / handler pairs. 
	 */
	private Hashtable<String,RequestHandler> requestHandlers;
	
	
	/**
	 * Hashtable of notification name / handler pairs.
	 */
	private Hashtable<String,NotificationHandler> notificationHandlers;
	
	
	/**
	 * Controls reporting of request processing time by appending a 
	 * non-standard "xProcTime" attribute to the JSON-RPC 2.0 response.
	 */
	private boolean reportProcTime = false;
	
	
	/**
	 * Creates a new dispatcher with no registered handlers.
	 */
	public Dispatcher() {
	
		requestHandlers = new Hashtable<String,RequestHandler>();
		notificationHandlers = new Hashtable<String,NotificationHandler>();
	}
	
	
	/**
	 * Registers a new JSON-RPC 2.0 request handler.
	 *
	 * @param handler The request handler to register.
	 *
	 * @throws IllegalArgumentException On attempting to register a handler
	 *                                  that duplicates an existing request
	 *                                  name.
	 */
	public void register(final RequestHandler handler) {
	
		for (String name: handler.handledRequests()) {
		
			if (requestHandlers.containsKey(name))
				throw new IllegalArgumentException("Cannot register a duplicate handler for request " + name);
		
			requestHandlers.put(name, handler);
		}
	}
	
	
	/**
	 * Registers a new JSON-RPC 2.0 notification handler.
	 *
	 * @param handler The notification handler to register.
	 *
	 * @throws IllegalArgumentException On attempting to register a handler
	 *                                  that duplicates an existing
	 *                                  notification name.
	 */
	public void register(final NotificationHandler handler) {
	
		for (String name: handler.handledNotifications()) {
		
			if (notificationHandlers.containsKey(name))
				throw new IllegalArgumentException("Cannot register a duplicate handler for notification " + name);
		
			notificationHandlers.put(name, handler);
		}
	}
	
	
	/**
	 * Returns the registered request names.
	 *
	 * @return The request names.
	 */
	public String[] handledRequests() {
	
		return requestHandlers.keySet().toArray(new String[0]);
	}
	
	
	/**
	 * Returns the registered notification names.
	 *
	 * @return The notification names.
	 */
	public String[] handledNotifications() {
	
		return notificationHandlers.keySet().toArray(new String[0]);
	}
	
	
	/**
	 * Gets the handler for the specified JSON-RPC 2.0 request name.
	 *
	 * @param requestName The request name to lookup.
	 *
	 * @return The corresponding request handler or {@code null} if none 
	 *         was found.
	 */
	public RequestHandler getRequestHandler(final String requestName) {
	
		return requestHandlers.get(requestName);
	}
	
	
	/**
	 * Gets the handler for the specified JSON-RPC 2.0 notification name.
	 *
	 * @param notificationName The notification name to lookup.
	 *
	 * @return The corresponding notification handler or {@code null} if
	 *         none was found.
	 */
	public NotificationHandler getNotificationHandler(final String notificationName) {
	
		return notificationHandlers.get(notificationName);
	}
	
	
	/**
	 * Dispatches the specified JSON-RPC 2.0 request to the appropriate 
	 * handler for processing and returns the response.
	 *
	 * @param request    The JSON-RPC 2.0 request to dispatch for 
	 *                   processing.
	 * @param requestCtx Context information about the request, may be 
	 *                   {@code null} if undefined.
	 *
	 * @return The response, which may indicate a processing error, such
	 *         as METHOD_NOT_FOUND.
	 */
	public JSONRPC2Response dispatch(final JSONRPC2Request request, final MessageContext requestCtx) {
	
		long startNanosec = 0;
		
		// Measure request processing time?
		if (reportProcTime)
			startNanosec = System.nanoTime();
		
	
		final String method = request.getMethod();
		
		RequestHandler handler = getRequestHandler(method);
		
		if (handler == null) {
		
			// We didn't find a handler for the requested RPC
		
			Object id = request.getID();
			
			return new JSONRPC2Response(JSONRPC2Error.METHOD_NOT_FOUND, id);
		}
			
		// Process the request
		
		JSONRPC2Response response = handler.process(request, requestCtx);
		
		if (reportProcTime) {
		
			final long procTimeNanosec = System.nanoTime() - startNanosec;
			
			response.appendNonStdAttribute("xProcTime", procTimeNanosec / 1000 + " us");
		}
		
		return response;
	}
	
	
	/**
	 * Dispatches the specified JSON-RPC 2.0 notification to the appropriate 
	 * handler for processing.
	 *
	 * <p>Note that JSON-RPC 2.0 notifications don't produce a response!
	 *
	 * @param notification    The JSON-RPC 2.0 notification to dispatch for 
	 *                        processing.
	 * @param notificationCtx Context information about the notification,
	 *                        may be {@code null} if undefined.
	 */
	public void dispatch(final JSONRPC2Notification notification, final MessageContext notificationCtx) {
	
		final String method = notification.getMethod();
		
		NotificationHandler handler = getNotificationHandler(method);
		
		if (handler == null) {
		
			// We didn't find a handler for the requested RPC
			return;
		}
			
		// Process the notification
		
		handler.process(notification, notificationCtx);
	}
	
	
	/**
	 * Controls reporting of request processing time by appending a 
	 * non-standard "xProcTime" attribute to the JSON-RPC 2.0 response.
	 * Reporting is disabled by default.
	 *
	 * @param enable {@code true} to enable proccessing time reporting, 
	 *               {@code false} to disable it.
	 */
	public void reportProcTime(final boolean enable) {
	
		reportProcTime = enable;
	}
	
	
	/**
	 * Returns {@code true} if reporting of request processing time is 
	 * enabled. See the {@link #reportProcTime} description for more
	 * information.
	 *
	 * @return {@code true} if reporting of request processing time is 
	 *         enabled, else {@code false}.
	 */
	public boolean reportsProcTime() {
	
		return reportProcTime;
	}
}
